Kernel/FileSystem/TmpFS.cpp

```diff
@@ -0,0 +1,295 @@
+{
+    LOCKER(m_lock);
+    ASSERT(identifier.fsid() == fsid());
+
+    auto it = m_inodes.find(identifier.index());
+    if (it == m_inodes.end())
+        return nullptr;
+    return it->value;
+}
+
+RefPtr<Inode> TmpFS::create_inode(InodeIdentifier parent_id, const String& name, mode_t mode, off_t size, dev_t dev, int& error)
+{
+    LOCKER(m_lock);
+    ASSERT(parent_id.fsid() == fsid());
+
+    ASSERT(size == 0);
+    ASSERT(dev == 0);
+
+    struct timeval now;
+    kgettimeofday(now);
+
+    InodeMetadata metadata {};
+    metadata.mode = mode;
+    metadata.uid = current->process().euid();
+    metadata.gid = current->process().egid();
+    metadata.atime = now.tv_sec;
+    metadata.ctime = now.tv_sec;
+    metadata.mtime = now.tv_sec;
+
+    auto inode = TmpFSInode::create(*this, metadata, parent_id);
+
+    auto it = m_inodes.find(parent_id.index());
+    ASSERT(it != m_inodes.end());
+    auto parent_inode = it->value;
+    error = parent_inode->add_child(inode->identifier(), name, mode);
+
+    return inode;
+}
+
+RefPtr<Inode> TmpFS::create_directory(InodeIdentifier parent_id, const String& name, mode_t mode, int& error)
+{
+    // Ensure it's a directory.
+    mode &= ~0170000;
+    mode |= 0040000;
+    return create_inode(parent_id, name, mode, 0, 0, error);
+}
+
+TmpFSInode::TmpFSInode(TmpFS& fs, InodeMetadata metadata, InodeIdentifier parent)
+    : Inode(fs, fs.next_inode_index())
+    , m_metadata(metadata)
+    , m_parent(parent)
+{
+    m_metadata.inode = identifier();
+}
+
+TmpFSInode::~TmpFSInode()
+{
+}
+
+NonnullRefPtr<TmpFSInode> TmpFSInode::create(TmpFS& fs, InodeMetadata metadata, InodeIdentifier parent)
+{
+    auto inode = adopt(*new TmpFSInode(fs, metadata, parent));
+    fs.register_inode(inode);
+    return inode;
+}
+
+NonnullRefPtr<TmpFSInode> TmpFSInode::create_root(TmpFS& fs)
+{
+    InodeMetadata metadata {};
+    metadata.mode = 0040777;
+    return create(fs, metadata, { fs.fsid(), 1 });
+}
+
+InodeMetadata TmpFSInode::metadata() const
+{
+    LOCKER(m_lock);
+
+    return m_metadata;
+}
+
+bool TmpFSInode::traverse_as_directory(Function<bool(const FS::DirectoryEntry&)> callback) const
+{
+    LOCKER(m_lock);
+
+    if (!is_directory())
+        return false;
+
+    for (auto& it : m_children)
+        callback(it.value.entry);
+    return true;
+}
+
+ssize_t TmpFSInode::read_bytes(off_t offset, ssize_t size, u8* buffer, FileDescription*) const
+{
+    LOCKER(m_lock);
+    ASSERT(!is_directory());
+
+    auto slice = m_content.slice_view(offset, size);
+    memcpy(buffer, slice.data(), slice.size());
+
+    return slice.size();
+}
+
+ssize_t TmpFSInode::write_bytes(off_t offset, ssize_t size, const u8* buffer, FileDescription*)
+{
+    LOCKER(m_lock);
+    ASSERT(!is_directory());
+
+    if (offset + size > m_content.size()) {
+        m_content.grow(offset + size);
+        m_metadata.size = offset + size;
+        set_metadata_dirty(true);
```

Comment:
What's going on here? :)

Reply:
There's a comment about it in `flush_metadata()`, basically this is for making sure we notify inode watchers (i.e. the file manager automatically refreshes). If/when inode watchers are switched to a more proper notification mechanism, we can stop doing this.